src/vcpkg/base/downloads.cpp

```diff
@@ -278,13 +280,35 @@ namespace vcpkg
         {
             cmd.string_arg(url);
         }
-        auto res = cmd_execute_and_stream_lines(cmd, [out](StringView line) {
+
+        std::vector<std::string> lines;
+
+        auto res = cmd_execute_and_stream_lines(cmd, [out, &lines](StringView line) {
+            lines.push_back(line.to_string());
             if (Strings::starts_with(line, guid_marker))
             {
                 out->push_back(std::strtol(line.data() + guid_marker.size(), nullptr, 10));
             }
         });
         Checks::check_exit(VCPKG_LINE_INFO, res == 0, "curl failed to execute with exit code: %d", res);
+
+        if (out->size() != start_size + urls.size())
+        {
+            // Invariant violation
+            print2(
+                Color::error,
+                "Error: vcpkg has encountered an internal invariant violation and cannot continue.\nPlease review the "
+                "following text for sensitive information and open an issue on the Microsoft/vcpkg GitHub to help fix "
+                "this problem!\n");
+
+            print2("cmd: ", cmd.command_line(), '\n');
+            print2("=== curl output ===\n");
+            for (auto&& line : lines)
+                print2(line, '\n');
+            print2("=== end curl output ===\n");
+
+            Checks::exit_fail(VCPKG_LINE_INFO);
```

Comment:
DECLARE_AND_REGISTER_MESSAGE(
        CurlReportedUnexpectedResults,
        (msg::command_line, msg::actual),
        "actual is the console output of curl.exe, expected to be program-to-program locale-invariant download results.",
        "Error: curl has reported unexpected results to vcpkg and vcpkg cannot continue.\n"
        "Please review the following text for sensitive information and open an issue on the Microsoft/vcpkg GitHub to help fix this problem!\n"
        "cmd: {command_line}\n"
        "=== curl output ===\n"
        "{actual}\n"
        "=== end curl output ===\n");
        Checks::msg_exit_fail(msgCurlReportedUnexpectedResults, msg::command_line = cmd.command_line(), msg::actual = Strings::join("\n", lines));
        

I think a solution like this would be permenant.